components/locid/src/helpers.rs

```diff
@@ -56,15 +56,150 @@ impl<T> From<Vec<T>> for ShortVec<T> {
+        $is_normalized:expr,
+        $error:ident,
+        [$good_example:literal $(,$more_good_examples:literal)*],
+        [$bad_example:literal $(, $more_bad_examples:literal)*],
+    ) => {
+        #[derive(Debug, PartialEq, Eq, Clone, Hash, PartialOrd, Ord, Copy)]
+        #[cfg_attr(feature = "serialize", derive(serde::Serialize))]
+        #[repr(transparent)]
+        $(#[$doc])*
+        pub struct $name(tinystr::TinyAsciiStr<$len_end>);
+
+        impl $name {
+            /// A constructor which takes a utf8 slice, parses it and
+            #[doc = concat!("produces a well-formed [`", stringify!($name), "`].")]
+            ///
+            /// # Examples
+            ///
+            /// ```
+            #[doc = concat!("use icu_locid::", stringify!($full_name), ";")]
+            ///
+            #[doc = concat!("let lang = ", stringify!($name), "::from_bytes(b", stringify!($good_example), ")")]
+            /// let lang = Language::from_bytes(b"en")
+            ///     .expect("Parsing failed.");
+            ///
+            #[doc = concat!("assert_eq!(lang, ", stringify!($good_example), ");")]
+            /// assert_eq!(lang, "en");
+            /// ```
+            pub const fn from_bytes(v: &[u8]) -> Result<Self, crate::parser::errors::ParserError> {
+                Self::from_bytes_manual_slice(v, 0, v.len())
+            }
+
+            /// Equivalent to [`from_bytes(bytes[start..end])`](Self::from_bytes),
+            /// but callable in a `const` context (which range indexing is not).
+            pub const fn from_bytes_manual_slice(
+                v: &[u8],
+                start: usize,
+                end: usize,
+            ) -> Result<Self, crate::parser::errors::ParserError> {
+                let slen = end - start;
+
+                #[allow(clippy::double_comparisons)] // if len_start == len_end
+                if slen < $len_start || slen > $len_end {
+                    return Err(crate::parser::errors::ParserError::$error);
+                }
+
+                match tinystr::TinyAsciiStr::from_bytes_manual_slice(v, start, end) {
+                    Ok(s) if s.len() >= $len_start && $condition(&s) => Ok(Self($normalization(s))),
+                    _ => Err(crate::parser::errors::ParserError::$error),
+                }
+            }
+
+            #[doc = concat!("Safely creates a [`", stringify!($name), "`] from a reference to its raw format")]
+            /// as returned by [`Self::into_raw()`]. Unlike the other constructors,
+            /// this one only takes normalized values.
+            pub fn try_from_raw(
+                v: [u8; $len_end],
+            ) -> Result<Self, crate::parser::errors::ParserError> {
+                let s = tinystr::TinyAsciiStr::<$len_end>::try_from_raw(v)
+                    .map_err(|_| crate::parser::errors::ParserError::$error)?;
+                if $len_start < s.len() && s.len() < $len_end && $is_normalized(&s) && $condition(&s) {
```

Comment:
`question`: your `TinyAsciiStr` is already bound to `len_end`. Do you need to test again here?

Reply:
you're right